### PR TITLE
feat!: total confidence

### PR DESCRIPTION
**Pull request by vahidlazio**

### CODE REVIEW

File: confidence/confidence.py [hunk 1]
```diff
@@ -69,66 +64,56 @@ class ResolveResult(object):
     token: str
 
 
-class ConfidenceOpenFeatureProvider(AbstractProvider):
+class Confidence:
+    context: Dict[str, str] = {}
```
Comment (nicklasl) [inline confidence/confidence.py:68]: I think we want/need to support multiple types.

File: confidence/confidence.py [hunk 2]
```diff
@@ -177,18 +152,46 @@ def _evaluate(
             flag_metadata={"flag_key": flag_key},
         )
 
-    def _resolve(self, flag_name: FlagName, context: Dict[str, str]) -> ResolveResult:
+    def track(self, event_name: str, context: Dict[str, str]) -> None:
```
Comment (nicklasl) [inline confidence/confidence.py:155]:
```suggestion
    def track(self, event_name: str, data: Dict[str, str]) -> None:
```
Also, data should be able to take more types..

File: confidence/confidence.py [hunk 3]
```diff
@@ -177,18 +152,46 @@ def _evaluate(
...
+                {
+                    "eventDefinition": f"eventDefinitions/{event_name}",
+                    "payload": {**self.context, **context},
+                    "eventTime": current_time,
+                }
+            ],
+            "sdk": {"id": "SDK_ID_PYTHON_CONFIDENCE", "version": __version__},
+        }
+
+        print(request_body)
```
Comment (nicklasl) [inline confidence/confidence.py:173]: remove.

File: confidence/confidence.py [hunk 4]
```diff
@@ -177,18 +152,46 @@ def _evaluate(
...
+
+        event_url = "https://events.confidence.dev/v1/events:publish"
+        headers = {"Content-Type": "application/json", "Accept": "application/json"}
+        response = requests.post(event_url, json=request_body, headers=headers)
+        response.raise_for_status()
```
Comment (nicklasl) [inline confidence/confidence.py:178]: where is this error handled?

File: confidence/errors.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import typing
+from enum import Enum
+
+
+class ErrorCode(Enum):
+    PROVIDER_NOT_READY = "NOT_READY"
```
Comment (nicklasl) [inline confidence/errors.py:6]:
```suggestion
    NOT_READY = "NOT_READY"
```

File: confidence/provider/provider.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    def resolve_boolean_details(
+        self,
+        flag_key: str,
+        default_value: bool,
+        evaluation_context: Optional[EvaluationContext] = None,
+    ) -> FlagResolutionDetails[bool]:
+        details = self._confidence_with_context(
+            evaluation_context
+        ).resolve_boolean_details(flag_key, default_value)
+        print("Confidence result", details)
```
Comment (nicklasl) [inline confidence/provider/provider.py:89]: clean up print

File: confidence/provider/provider.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    def resolve_string_details(
+        self,
+        flag_key: str,
+        default_value: str,
+        evaluation_context: Optional[EvaluationContext] = None,
+    ) -> FlagResolutionDetails[str]:
+        details = self._confidence_with_context(
+            evaluation_context
+        ).resolve_string_details(flag_key, default_value)
+        print("Confidence result", details)
```
Comment (nicklasl) [inline confidence/provider/provider.py:144]: clean up print

File: tests/test_provider.py [hunk 8]
```diff
@@ -28,6 +32,8 @@ def test_resolve_string_with_dot_notation(self):
                 evaluation_context=ctx,
             )
 
+            print("result", result)
```
Comment (nicklasl) [inline tests/test_provider.py:35]: clean up print

File: confidence/confidence.py (hunk 4 above)
Comment (nicklasl) [reply confidence/confidence.py:178]: I think we need to figure out how to handle the response using some logging to signal if things are not being accepted.

File: demo.py [hunk 9]
```diff
@@ -9,16 +9,18 @@ async def get_flag():
...
     value = confidence.resolve_string_details("hawkflag.color", "False")
     print(f"Flag value: {value}")
 
 
 # Another asynchronous function that calls the first one
 async def main():
     await get_flag()
-    print("Finished calling say_hello again")
+    print("Finished calling get_flag")
+    await asyncio.sleep(1)
```
Comment (vahidlazio) [inline demo.py:22]: why do we sleep here?

**Review by nicklasl — APPROVED** [review]

File: demo.py (hunk 9 above)
Comment (nicklasl) [reply demo.py:22]: good point, let's make that.

**Pull request merged**